add kirenenko

chenju2k6 opened the pull request:
Signed-off-by: Ju Chen <chenju2k6@gmail.com>

chenju2k6 commented:
Currently, Kirenenko can support 7 benchmarks 
- libpng
- libjpeg
- lcms
- libxml2
- sqlite3
- vorbis
- mbedtls

lszekeres commented on fuzzers/kirenenko/buildScript/lcms-2017-03-21/buildK.sh:
```diff
@@ -0,0 +1,25 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Can we avoid these redundant build scripts, and use the original `benchmarks/*/build.sh`-s everywhere?

To build the two binaries, consider calling the original build script twice from the fuzzer.py's `build()`, similarly as you're doing it now, with different env vars. To avoid the name conflict, perhaps we could add the `_kir` and `_vani` suffixes by renaming the output file after calling `utils.build_benchmark()` in `build()`?

Feel free to make necessary modifications to the original build scripts when necessary (e.g., add missing `make clean`, or change CXX to CC, use `cp -rf`, etc). That way we won't need three very similar build scripts for each benchmark, only one.

chenju2k6 replied:
Thanks for the note. Are we going to directly modify the build script in each benchmarks directory? Would that impact the other fuzzers and measures? Or do we need to modify the script in build()?

lszekeres replied:
> Are we going to directly modify the build script in each benchmarks directory? 

Yes, please.

> Would that impact the other fuzzers and measures? 

Yes, all fuzzers should use the same `build.sh`. Let us know if that causes issues.

chenju2k6 replied:
Thanks for the note. I think I've fixed the requested changes in the new commit.

lszekeres commented on fuzzers/kirenenko/buildScript/lcms-2017-03-21/cms_transform_fuzzer.c:
```diff
@@ -0,0 +1,61 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdint.h>
```
For the fuzz target source files as well, can we avoid the redundant copies?

It looks like the changes are turning the c++ files into c. Please feel free to turn the original .cc files into .c file and use the C compiler everywhere, and that way we don't need these very similar copies. Would that work?

chenju2k6 replied:
Got it. In addition to turning the filename, we also convert some C++ syntax to its C equivalent. (e.g. new -> malloc). I will try to fix it in the script.

lszekeres replied:
> Got it. In addition to turning the filename, we also convert some C++ syntax to its C equivalent. (e.g. new -> malloc). I will try to fix it in the script.

Yes, please feel free to make these changes in the fuzz target source files and turn the C++ bits to C in the original files. It's important that we build the same fuzz target source for all fuzzers.

jonathanmetzman replied:
> It looks like the changes are turning the c++ files into c. Please feel free to turn the original .cc files into .c file and use the C compiler everywhere, and that way we don't need these very similar copies. Would that work?

In OSS-Fuzz we advocate using clang++ to link things. I'm not 100% sure why and I don't know if linking with `clang` will fail. but I'm quite sure some of these projects are written in C++. So what's the point of changing it here?

jonathanmetzman replied:
> > It looks like the changes are turning the c++ files into c. Please feel free to turn the original .cc files into .c file and use the C compiler everywhere, and that way we don't need these very similar copies. Would that work?
> 
> In OSS-Fuzz we advocate using clang++ to link things. I'm not 100% sure why and I don't know if linking with `clang` will fail. but I'm quite sure some of these projects are written in C++. So what's the point of changing it here?

Ah, I bet kirienko only supports pure C projects right?

chenju2k6 replied:
Yes, currently Kirenenko only supports C projects.

lszekeres reviewed (changes requested):
Thanks a lot!

jonathanmetzman commented on benchmarks/libjpeg-turbo-07-2017/libjpeg_turbo_fuzzer.c:
```diff
@@ -35,11 +35,13 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+        jpegDecompressor, data, size, buf, width, 0, height, TJPF_RGB, 0);
 
     tjDestroy(jpegDecompressor);
+		free(buf);
```
nit: fix this indentation please.

jonathanmetzman commented on benchmarks/libjpeg-turbo-07-2017/libjpeg_turbo_fuzzer.c:
```diff
@@ -35,11 +35,13 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
     }
 
-    std::unique_ptr<unsigned char[]> buf(new unsigned char[width * height * 3]);
+    //std::unique_ptr<unsigned char[]> buf(new unsigned char[width * height * 3]);
```
nit: delete this line please.

jonathanmetzman commented on benchmarks/libxml2-v2.9.2/target.c:
```diff
@@ -12,18 +12,18 @@
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
   xmlSetGenericErrorFunc(NULL, &ignore);
-  if (auto doc = xmlReadMemory(reinterpret_cast<const char *>(data), size,
+	xmlDocPtr doc;
```
nit: fix alignment please.

jonathanmetzman commented on fuzzers/kirenenko/builder.Dockerfile:
```diff
@@ -0,0 +1,78 @@
+
+ARG parent_image=gcr.io/fuzzbench/base-builder
+FROM $parent_image
+#FROM ubuntu:16.04
```
nit: delete thsi line please.

jonathanmetzman commented on benchmarks/libpng-1.2.56/target.c:
```diff
@@ -36,7 +36,7 @@ void user_read_data(png_structp png_ptr, png_bytep data, png_size_t length) {
 
 static const int kPngHeaderSize = 8;
-
+/*
```
nit: just delete the commented out code.

jonathanmetzman commented on fuzzers/kirenenko/builder.Dockerfile:
```diff
@@ -0,0 +1,78 @@
+                       apt-utils apt-transport-https ca-certificates
+
+
+#download install AFL++
```
nit: we begin comments with a capital letter and end them with a punctuation mark.
Please change to "# Download and install AFL++."

jonathanmetzman commented on fuzzers/kirenenko/builder.Dockerfile:
```diff
@@ -0,0 +1,78 @@
+RUN cd /hiredis && make -j && make install
+
+
+#RUN git clone https://github.com/sewenew/redis-plus-plus.git /redis-plus-plus
```
nit: Just delete these lines.

jonathanmetzman commented on fuzzers/kirenenko/builder.Dockerfile:
```diff
@@ -0,0 +1,78 @@
+RUN git clone https://github.com/ChengyuSong/Kirenenko.git  /Kirenenko
+COPY kir.patch /Kirenenko/kir.patch
+RUN cd /Kirenenko && git checkout 0390f27d23cd69337a1e8f31cd5cc93422107ec2  && patch -p1 < kir.patch && ./build/build.sh
+#COPY kirenenko-pro /Kirenenko
```
Nit: Delete these please.

jonathanmetzman commented on fuzzers/kirenenko/fuzzer.py:
```diff
@@ -0,0 +1,200 @@
+
+
+def is_benchmark(name):
+    "check if benchmark matches"
```
nit: We begin docstrings with triple quotes.
So please change this to:
```
"""Checks if benchmark matches."""
```

jonathanmetzman commented on fuzzers/kirenenko/fuzzer.py:
```diff
@@ -0,0 +1,200 @@
+    os.environ['AFL_LLVM_DICT2FILE'] = build_directory + '/afl++.dict'
+    env = os.environ.copy()
+    build_script = os.path.join(os.environ['SRC'], 'build.sh')
+    print("build script path is " + build_script)
```
nit: Other than for docstrings (super confusing and arbitrary) we use single quotes for strings, so please change to 
`print('build script path is ' + build_script)`

lszekeres commented on benchmarks/mbedtls_fuzz_dtlsclient/build.sh:
```diff
@@ -17,13 +17,18 @@
+if [ ! -d crypto ]; then
+    git -C crypto checkout -f 819799cfc68e4c4381673a8a27af19802c8263f2
+fi
+if [ ! -d build ]; then
+    mkdir build
+fi
```
consider dropping the if for simplicity and use `mkdir -p build`

lszekeres commented on benchmarks/mbedtls_fuzz_dtlsclient/build.sh:
```diff
@@ -17,13 +17,18 @@
 perl scripts/config.pl set MBEDTLS_PLATFORM_TIME_ALT
-git -C crypto checkout -f 819799cfc68e4c4381673a8a27af19802c8263f2
-mkdir build
+if [ ! -d crypto ]; then
```
consider dropping the `if` for simplicity, it's fine to run checkout again

lszekeres commented on fuzzers/kirenenko/discard.list:
```diff
@@ -0,0 +1,93 @@
+fun:jsimd_can_rgb_ycc=discard
```
Another, simpler, option would be avoiding asm code all together, so we wouldn't need this discard list for dfsan. For libjpeg-turbo, we can do that with `-DWITH_SIMD=0`. https://github.com/libjpeg-turbo/libjpeg-turbo#memory-debugger-pitfalls

One question whether we should use that flag (1) with all fuzzers ie. in the libjpeg build script or (2) let fuzzer.py config's decide to add the flag or not.

I'd recommend (2), because it'd allow using MSAN with all fuzzers (MSAN also doens't work well with asm). Wdyt @jonathanmetzman ?

lszekeres reviewed (changes requested):
Thank you! 

The CI builds are failing (see Checks tab). Could you please take a look?

chenju2k6 commented:
Looks like it is because that afl_driver.cpp can't compile with $CC. Let me fix it

lszekeres commented:
> Looks like it is because that afl_driver.cpp can't compile with $CC. Let me fix it

Friendly ping. Did you have a chance to look into this?

chenju2k6 commented:
> > Looks like it is because that afl_driver.cpp can't compile with $CC. Let me fix it
> 
> Friendly ping. Did you have a chance to look into this?

Sorry for the slow follow-up of this issue. Let me get this done tomorrow.

chenju2k6 commented:
> > Looks like it is because that afl_driver.cpp can't compile with $CC. Let me fix it
> 
> Friendly ping. Did you have a chance to look into this?

Just realized that 

> > Looks like it is because that afl_driver.cpp can't compile with $CC. Let me fix it
> 
> Friendly ping. Did you have a chance to look into this?

I may have to include the below ugly fix to solve this (just because Kirenenko currently can't compile C++). Is something like this acceptable? If not, we are working on C++ programs, and we then have a better fix if it is done.

`+#ifdef KIRENENKO
 int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+#else
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+#endif`

lszekeres commented:
> I may have to include the below ugly fix to solve this (just because Kirenenko currently can't compile C++). Is something like this acceptable? If not, we are working on C++ programs, and we then have a better fix if it is done.

That's fine, but consider using something more generic, e.g.:

```
#ifdef __cplusplus
extern “C”
#endif
int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { ... }
```

chenju2k6 commented:
> > I may have to include the below ugly fix to solve this (just because Kirenenko currently can't compile C++). Is something like this acceptable? If not, we are working on C++ programs, and we then have a better fix if it is done.
> 
> That's fine, but consider using something more generic, e.g.:
> 
> ```
> #ifdef __cplusplus
> extern “C”
> #endif
> int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { ... }
> ```

Thanks. Looks like my commit fails the checks again.  For the failed coverage builds such as **build (coverage, oss-fuzz)**, may I know how to reproduce the issue locally?  For the failed benchmark building for Kirenenko (lcms), may I know how to reproduce the issue locally?  In my local environment, I've tried "make build-kirenenko-lcms-2017-03-21", seems it runs fine.

inferno-chromium commented:
> > > I may have to include the below ugly fix to solve this (just because Kirenenko currently can't compile C++). Is something like this acceptable? If not, we are working on C++ programs, and we then have a better fix if it is done.
> > 
> > 
> > That's fine, but consider using something more generic, e.g.:
> > ```
> > #ifdef __cplusplus
> > extern “C”
> > #endif
> > int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { ... }
> > ```
> 
> Thanks. Looks like my commit fails the checks again. For the failed coverage builds such as **build (coverage, oss-fuzz)**, may I know how to reproduce the issue locally? For the failed benchmark building for Kirenenko (lcms), may I know how to reproduce the issue locally? In my local environment, I've tried "make build-kirenenko-lcms-2017-03-21", seems it runs fine.

make build-coverage-<fuzzer_name> to reproduce coverage build failures locally.

chenju2k6 commented:
> > > > I may have to include the below ugly fix to solve this (just because Kirenenko currently can't compile C++). Is something like this acceptable? If not, we are working on C++ programs, and we then have a better fix if it is done.
> > > 
> > > 
> > > That's fine, but consider using something more generic, e.g.:
> > > ```
> > > #ifdef __cplusplus
> > > extern “C”
> > > #endif
> > > int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { ... }
> > > ```
> > 
> > 
> > Thanks. Looks like my commit fails the checks again. For the failed coverage builds such as **build (coverage, oss-fuzz)**, may I know how to reproduce the issue locally? For the failed benchmark building for Kirenenko (lcms), may I know how to reproduce the issue locally? In my local environment, I've tried "make build-kirenenko-lcms-2017-03-21", seems it runs fine.
> 
> make build-coverage-<fuzzer_name> to reproduce coverage build failures locally.

Guys, I have tried the below commands locally, seems they are running fine. Could you help me to figure out why the patch can't pass the auto-build checks?

make-coverage-all
make-run-<benchmark_name>,  with 7 benchmarks, lcms, libxml2, libpng, libjpeg, mbedtls, vorbis and sqlite

jonathanmetzman commented:
> > > > > I may have to include the below ugly fix to solve this (just because Kirenenko currently can't compile C++). Is something like this acceptable? If not, we are working on C++ programs, and we then have a better fix if it is done.
> > > > 
> > > > 
> > > > That's fine, but consider using something more generic, e.g.:
> > > > ```
> > > > #ifdef __cplusplus
> > > > extern “C”
> > > > #endif
> > > > int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { ... }
> > > > ```
> > > 
> > > 
> > > Thanks. Looks like my commit fails the checks again. For the failed coverage builds such as **build (coverage, oss-fuzz)**, may I know how to reproduce the issue locally? For the failed benchmark building for Kirenenko (lcms), may I know how to reproduce the issue locally? In my local environment, I've tried "make build-kirenenko-lcms-2017-03-21", seems it runs fine.
> > 
> > 
> > make build-coverage-<fuzzer_name> to reproduce coverage build failures locally.
> 
> Guys, I have tried the below commands locally, seems they are running fine. Could you help me to figure out why the patch can't pass the auto-build checks?
> 
> make-coverage-all
> make-run-<benchmark_name>, with 7 benchmarks, lcms, libxml2, libpng, libjpeg, mbedtls, vorbis and sqlite

Sorry for the delayed response (a lot of people are out due to holidays). I'll take a look now.

jonathanmetzman commented:
>Guys, I have tried the below commands locally, seems they are running fine. Could you help me to figure out why the patch can't pass the auto-build checks?

Coverage failures seem to be a bug on our end. Sorry!
If you pull again and push, it should no longer fail.

The failure on "(kirenenko, oss-fuzz)" seems to be because of a timeout. I don't think we can do anything there. So if youre confident that all of the benchmarks that build does work, we can merge this PR without that completing.
The failure on "(kirenenko, standard)" is in lcms. Here is the line where it fails:
```
home/runner/work/_temp/9b757658-0a0c-4fc6-829a-8890c5d1cac9.sh: line 1:  2717 Killed                  PATH=.venv/bin/:$PATH PYTHONPATH=. python3 .github/workflows/build_and_test_run_fuzzer_benchmarks.py standard kirenenko
make: *** [.kirenenko-lcms-2017-03-21-intermediate-runner] Killed
```

That makes me think building runs out of memory. I think in production we build on better machines.
I'll rerun the builds and see if this fails again. If it does I guess we can merge. But it might be better to make the build use less threads so the build is less likely to run out of memory (I assume the OOM killer did this).

jonathanmetzman commented:
> The failure on "(kirenenko, standard)" is in lcms. Here is the line where it fails:
> 
> ```
> home/runner/work/_temp/9b757658-0a0c-4fc6-829a-8890c5d1cac9.sh: line 1:  2717 Killed                  PATH=.venv/bin/:$PATH PYTHONPATH=. python3 .github/workflows/build_and_test_run_fuzzer_benchmarks.py standard kirenenko
> make: *** [.kirenenko-lcms-2017-03-21-intermediate-runner] Killed
> ```
> 
> That makes me think building runs out of memory. I think in production we build on better machines.
> I'll rerun the builds and see if this fails again. If it does I guess we can merge. But it might be better to make the build use less threads so the build is less likely to run out of memory (I assume the OOM killer did this).

Because I reran the build you can't see the logs again until the build finishes. But I think the build was killed when Z3 was being built. Which is weird because Z3 is built in the OSS-Fuzz build and didn't seem to cause an OOM.

chenju2k6 commented:
> > Guys, I have tried the below commands locally, seems they are running fine. Could you help me to figure out why the patch can't pass the auto-build checks?
> 
> Coverage failures seem to be a bug on our end. Sorry!
> If you pull again and push, it should no longer fail. I think I've fixed it (if not just ignore please).
> 
> The failure on "(kirenenko, oss-fuzz)" seems to be because of a timeout. I don't think we can do anything there. So if youre confident that all of the benchmarks that build does work, we can merge this PR without that completing.
> The failure on "(kirenenko, standard)" is in lcms. Here is the line where it fails:
> 
> ```
> home/runner/work/_temp/9b757658-0a0c-4fc6-829a-8890c5d1cac9.sh: line 1:  2717 Killed                  PATH=.venv/bin/:$PATH PYTHONPATH=. python3 .github/workflows/build_and_test_run_fuzzer_benchmarks.py standard kirenenko
> make: *** [.kirenenko-lcms-2017-03-21-intermediate-runner] Killed
> ```
> 
> That makes me think building runs out of memory. I think in production we build on better machines.
> I'll rerun the builds and see if this fails again. If it does I guess we can merge. But it might be better to make the build use less threads so the build is less likely to run out of memory (I assume the OOM killer did this).

Hi. I replaced "make -j" with "make" to reduce resource usage for the building.  Also I tried "make debug-kirenenko-lcms-2017-03-21" which runs fine. I hope there will be no crash anymore.

jonathanmetzman reviewed (approved):
LGTM. I'll merge this Monday morning instead of now to avoid a weekend breakage.

jonathanmetzman reviewed (approved):
LGTM

jonathanmetzman commented:
Sorry, I'm going to have to revert this.
The if kirenenko in each build.sh is pretty ugly stuff and I don't like how it makes our code complex.
I can see why you did it. It's because fuzzers like AFL need to be compiled with $CXX but kirenenko can't use it.
I'm going to try to think of a solution.

chenju2k6 commented:
Thank you for mentioning this. We are addressing the $CXX problem.




On Mon, Jan 11, 2021 at 10:12 AM jonathanmetzman <notifications@github.com>
wrote:

> Sorry, I'm going to have to revert this.
> The if kirenenko in each build.sh is pretty ugly stuff and I don't like
> how it makes our code complex.
> I can see why you did it. It's because fuzzers like AFL need to be
> compiled with $CXX but kirenenko can't use it.
> I'm going to try to think of a solution.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/fuzzbench/pull/921#issuecomment-758129334>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXYIZDWQZ6HTY76PTBB73LSZM5RRANCNFSM4TEPJRCA>
> .
>

jonathanmetzman commented:
> Thank you for mentioning this. We are addressing the $CXX problem.

Thanks. Really sorry for this confusion.

jonathanmetzman commented:
> > Thank you for mentioning this. We are addressing the $CXX problem.
> 
> Thanks. Really sorry for this confusion.

I think I can get the C version of libpng to compile with afl if AFL driver is compiled with lc++. This is one option but I don't think it's a good one (since it may need to be applied to AFL++ and other variants as well). I think the best solution would be for kirenenko to support CXX
